src/main/java/com/google/research/bleth/servlets/EnqueueExperimentServlet.java

```diff
@@ -0,0 +1,234 @@
+ * in order to create and run a new experiment.
+ */
+@WebServlet("/enqueue-experiment")
+public class EnqueueExperimentServlet extends HttpServlet {
+    private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    private static final Gson gson = new Gson();
+    private static final MovementStrategyFactory.Type defaultMovementStrategy = MovementStrategyFactory.Type.RANDOM;
+    private static final AwakenessStrategyFactory.Type defaultAwakenessStrategy = AwakenessStrategyFactory.Type.RANDOM;
+    private static final String PROJECT_ID = "bleth-2020";
+    private static final String LOCATION_ID = "europe-west1";
+    private static final String QUEUE_ID = "simulations-queue";
+    private static final String queueName = QueueName.of(PROJECT_ID, LOCATION_ID, QUEUE_ID).toString();
+    private static final Logger log = Logger.getLogger(EnqueueExperimentServlet.class.getName());
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        if (request.getServerName().equals("localhost") && request.getServerPort() == 8080) {
+            response.setContentType("text/plain;");
+            response.getWriter().println("Experiments are not yet supported on localhost.");
+            return;
+        }
+
+        log.info("Enqueue Experiment Servlet invoked.");
+
+        String experimentTitle = request.getParameter("experimentTitle");
+        int beaconsNum = Integer.parseInt(request.getParameter("beaconsNum"));
+        Set<List<PropertyWrapper>> configurations = createConfigurations(request);
+
+        Entity experiment = new Entity(Schema.Experiment.entityKind);
+        experiment.setProperty(Schema.Experiment.experimentTitle, experimentTitle);
+        Key experimentId = datastore.put(experiment);
+        log.info("A new experiment entity with id " + KeyFactory.keyToString(experimentId) +
+                "was created and written to db.");
+
+        int legalConfigurationsCount = configurations.size();
+        for (List<PropertyWrapper> configuration : configurations) {
+            AppEngineHttpRequest httpRequest = toHttpRequest(configuration, beaconsNum, experimentId, experimentTitle);
+            log.info("A new AppEngineHttpRequest was created: " + httpRequest.toString());
+            enqueueTask(httpRequest);
+            log.info("A new Task was created and enqueued.");
+        }
+
+        try {
+            experiment = datastore.get(experimentId);
+            experiment.setProperty(Schema.Experiment.simulationsLeft, legalConfigurationsCount);
+            datastore.put(experiment);
+            log.info("Experiment entity with id " + KeyFactory.keyToString(experimentId) +
+                    "was updated with simulationsLeft=" + legalConfigurationsCount);
+        } catch (EntityNotFoundException e) {
+            response.setContentType("text/plain;");
+            response.getWriter().println(e.getMessage());
```

Comment:
I actually think we can add the entity's `simulationsLeft` property while it is being created (right after line 85) instead of having this try-catch block (the reason this try-catch block exists is that previously the arguments validation occurred inside the for loop, and this number of legal configs was known only after the for loop - this is not the case anymore). 

What do you think?
@nolaviz @OrpazFishel

Reply:
SG